contrib/transfer_hierarchy.lua

```diff
@@ -0,0 +1,347 @@
+    local sourceBase = sourceTextBox.text
+    local destBase = destinationTextBox.text
+    local destFilms = {}
+    for film, _ in pairs(films) do
+       films[film] = destBase .. string.sub(film.path, #sourceBase+1)
+       if not pathExists(films[film]) then
+           if createDirectory(films[film]) == nil then
+	       darktable.print("transfer hierarchy: ERROR: could not create directory: " .. films[film])
+	       return
+	   end
+       end
+       if not pathIsDirectory(films[film]) then
+           darktable.print("transfer hierarchy: ERROR: not a directory: " .. films[film])
+	   return
+       end
+       destFilms[film] = darktable.films.new(films[film])
+       if destFilms[film] == nil then
+           darktable.print("transfer hierarchy: ERROR: could not create film: " .. film.path)
+       end
+    end
+
+    local job = darktable.gui.create_job(string.format("transfer hierarchy (%d image" .. (#(darktable.gui.action_images) == 1 and "" or "s") .. ")", #(darktable.gui.action_images)), true, stopTransfer)
+    job.percent = 0.0
+    local pctIncrement = 1.0 / #(darktable.gui.action_images)
+    for _,img in ipairs(darktable.gui.action_images) do
+       if job.valid then
+    	  destFilm = destFilms[img.film]
+	  transferFunc(img, destFilm)
+	  job.percent = job.percent + pctIncrement
+       end
+    end
+    job.valid = false
+    local filterRules = darktable.gui.libs.collect.filter()
+    darktable.gui.libs.collect.filter(filterRules)
+end
+
+local function doMove()
+    doTransfer(darktable.database.move_image)
+end
+
+local function doCopy()
+    doTransfer(darktable.database.copy_image)
+end
+
+
+
+
+
+
+local transfer_widget = darktable.new_widget("box") {
+   orientation = "vertical",
+   darktable.new_widget("button") {
+     label = "calculate",
+     clicked_callback = doCalculate
+   },
+   darktable.new_widget("label") {
+      label = "existing root",
+      halign = "start"
+   },
+   sourceTextBox,
+   darktable.new_widget("label") {
+     label = "root of destination",
+     halign = "start"
+   },
+   destinationTextBox,
+   darktable.new_widget("button") {
+     label = "move",
+     tooltip = "Move all selected images",
+     clicked_callback = doMove
+   },
+   darktable.new_widget("button") {
+     label = "copy",
+     tooltip = "Copy all selected images",
+     clicked_callback = doCopy
+   }
+}
+
+-- Widgets and business logic: END
+
+
+
+
+
+
+-- Preferences: BEGIN
+
+darktable.preferences.register(
+      LIB_ID,
+      "source_base",
+      "string",
+      "[transfer hierarchy] Customary source root",
+      "",
+      "")
+
+darktable.preferences.register(
+      LIB_ID,
+      "destination_base",
+      "string",
+      "[transfer hierarchy] Customary destination root",
+      "",
+      "")
+
+-- Preferences: END
+
+
+
+
+
+
+darktable.register_lib(LIB_ID,
+		       "transfer hierarchy", true, true, {
+			  [darktable.gui.views.lighttable] = { "DT_UI_CONTAINER_PANEL_RIGHT_CENTER", 1000 }
```

Comment:
Setting the position to 700 puts transfer_hierarchy just below select, and selected_images.  1000 puts it at the top, above select, which might come as a surprise to users.

Reply:
Done in commit 0501758.